File: bookworm/lib/model/Bill.dart
class Bill {
  final String id;
  final String requestId;
  final String type;
  final int? overdueDays;
  final int? overdueFee; // VND
  final double? damageFee; // VND
  final double totalFee; // VND
  final double amountReceived; // VND khách đưa
  final double changeGiven; // VND trả lại
  final DateTime date; // ngày lập hóa đơn

  Bill({
    required this.id,
    required this.requestId,
    required this.type,
    this.overdueDays,
    this.overdueFee,
    this.damageFee,
    required this.totalFee,
    required this.amountReceived,
    required this.changeGiven,
    DateTime? date,
  }) : date = date ?? DateTime.now();

  factory Bill.fromJson(Map<String, dynamic> json) =>
      Bill(
        id: json['id'],
        requestId: json['borrowRequestId'],
        // thay json['request_id']
        type: json['type'],
        overdueDays: json['overdueDays'] as int?,
        overdueFee: json['overdueFee'] as int?,
        damageFee: (json['damageFee'] as num?)?.toDouble(),
        totalFee: (json['totalFee'] as num).toDouble(),
        amountReceived: (json['amountReceived'] as num).toDouble(),
        changeGiven: (json['changeGiven'] as num).toDouble(),
        date: DateTime.parse(json['date']),
      );

  Map<String, dynamic> toJson() =>
      {
        'id': id,
        'borrowRequestId': requestId, // thay request_id
        'type': type,
        'overdueDays': overdueDays,
        'overdueFee': overdueFee,
        'damageFee': damageFee,
        'totalFee': totalFee,
        'amountReceived': amountReceived,
        'changeGiven': changeGiven,
        'date': date.toIso8601String(),
      };
}

File: bookworm/service/RequestService/routes/bill.js
const express = require('express');
const router = express.Router();
const Bill = require('../models/Bill');

/**
 * @swagger
 * /api/bill:
 *   get:
 *     summary: Lấy danh sách tất cả các hóa đơn
 *     tags: [Bills]
 *     responses:
 *       200:
 *         description: Danh sách hóa đơn
 *       500:
 *         description: Lỗi khi truy vấn hóa đơn
 */
// lấy danh sách tất cả các hóa đơn
router.get('/', async (req, res) => {
  try {
    const bills = await Bill.find().sort({ date: -1 });
    res.json(bills);
  } catch (err) {
    res.status(500).json({ error: 'Failed to fetch bills', details: err.message });
  }
});

/**
 * @swagger
 * /api/bill:
 *   post:
 *     summary: Tạo mới một hóa đơn
 *     tags: [Bills]
 *     requestBody:
 *       required: true
 *       content:
 *         application/json:
 *           schema:
 *             type: object
 *             properties:
 *               borrowRequestId:
 *                 type: string
 *               type:
 *                 type: string
 *                 enum: [book, room]
 *               overdueDays:
 *                 type: number
 *               overdueFee:
 *                 type: number
 *               damageFee:
 *                 type: number
 *               totalFee:
 *                 type: number
 *               amountReceived:
 *                 type: number
 *               changeGiven:
 *                 type: number
 *               date:
 *                 type: string
 *                 format: date-time
 *     responses:
 *       201:
 *         description: Hóa đơn được tạo thành công
 *       400:
 *         description: Dữ liệu không hợp lệ hoặc lỗi khi tạo
 */
// tạo mới một hóa đơn
router.post('/', async (req, res) => {
  try {
    const bill = new Bill(req.body);
    await bill.save();
    res.status(201).json(bill);
  } catch (err) {
    res.status(400).json({ error: 'Failed to create bill', details: err.message });
  }
});

module.exports = router;
